From 42f0e65be2b305f817ca8b93369d1af6829075c8 Mon Sep 17 00:00:00 2001
From: Mohammed Hassan <mohammed.hassan@jollamobile.com>
Date: Sat, 1 Dec 2012 06:17:10 +0200
Subject: [PATCH 14/18] Add libresourceqt1 to PKGCONFIG when building for
 MeeGo

---
 plugins/multimedia/gstreamer/gstreamer.pro |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/plugins/multimedia/gstreamer/gstreamer.pro b/plugins/multimedia/gstreamer/gstreamer.pro
index bbd59a0..d3c7661 100644
--- a/plugins/multimedia/gstreamer/gstreamer.pro
+++ b/plugins/multimedia/gstreamer/gstreamer.pro
@@ -38,6 +38,10 @@ maemo5 {
   QT += dbus
 }
 
+meego {
+    PKGCONFIG += libresourceqt1
+}
+
 maemo6 {
     HEADERS += camerabuttonlistener_meego.h
     SOURCES += camerabuttonlistener_meego.cpp
-- 
1.7.10.3

