Index: qt-mobility-opensource-src-1.2.0/src/serviceframework/ipc/qmetaobjectbuilder_p.h
===================================================================
--- qt-mobility-opensource-src-1.2.0.orig/src/serviceframework/ipc/qmetaobjectbuilder_p.h
+++ qt-mobility-opensource-src-1.2.0/src/serviceframework/ipc/qmetaobjectbuilder_p.h
@@ -172,7 +172,7 @@ public:
     int indexOfEnumerator(const QByteArray& name);
     int indexOfClassInfo(const QByteArray& name);
 
-    typedef int (*StaticMetacallFunction)(QMetaObject::Call, int, void **);
+    typedef QMetaObjectExtraData::StaticMetacallFunction StaticMetacallFunction;
 
     QMetaObjectBuilder::StaticMetacallFunction staticMetacallFunction() const;
     void setStaticMetacallFunction(QMetaObjectBuilder::StaticMetacallFunction value);
