Index: qt-mobility-opensource-src-1.2.0/src/messaging/qmessageglobal.h
===================================================================
--- qt-mobility-opensource-src-1.2.0.orig/src/messaging/qmessageglobal.h
+++ qt-mobility-opensource-src-1.2.0/src/messaging/qmessageglobal.h
@@ -59,7 +59,8 @@
     }
 #  else
 #  define Q_SCOPED_STATIC_DEFINE(TYPE, SCOPE, NAME)                     \
-    Q_GLOBAL_STATIC_INIT(TYPE, NAME);                                   \
+    static QGlobalStatic<TYPE > this_##NAME                             \
+                = { Q_BASIC_ATOMIC_INITIALIZER(0), false };             \
     TYPE *SCOPE::NAME()                                                 \
     {                                                                   \
         if (!this_##NAME.pointer && !this_##NAME.destroyed) {           \
Index: qt-mobility-opensource-src-1.2.0/src/bearer/qnetworkconfigmanager.cpp
===================================================================
--- qt-mobility-opensource-src-1.2.0.orig/src/bearer/qnetworkconfigmanager.cpp
+++ qt-mobility-opensource-src-1.2.0/src/bearer/qnetworkconfigmanager.cpp
@@ -76,7 +76,8 @@ QTM_BEGIN_NAMESPACE
     }
 #else
 #define Q_GLOBAL_STATIC_QAPP_DESTRUCTION(TYPE, NAME)                    \
-    Q_GLOBAL_STATIC_INIT(TYPE, NAME);                                   \
+    static QGlobalStatic<TYPE > this_##NAME                             \
+                = { Q_BASIC_ATOMIC_INITIALIZER(0), false };             \
     static void NAME##_cleanup()                                        \
     {                                                                   \
         delete this_##NAME.pointer;                                     \
